query-algebrizer/src/clauses/mod.rs

```diff
@@ -825,7 +826,139 @@ impl ConjoiningClauses {
     /// where `?y` must simultaneously be a ref-typed attribute and a boolean-typed attribute. This
     /// function deduces that and calls `self.mark_known_empty`. #293.
     pub fn expand_type_tags(&mut self) {
-        // TODO.
+        for (v, tag) in self.extracted_types.iter() {
+            let cols = self.column_bindings.get(v).unwrap().clone();
+            for col in cols {
+                if col.0 != tag.0 { 
+                    let value = QueryValue::Column(QualifiedAlias(col.0.clone(), Column::Fixed(DatomsColumn::ValueTypeTag)));
+                    let alias = tag.clone();
+                    self.wheres.0.push(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(alias, value)));
+                }
+            }
+        }
+
+        let mut computed_types: BTreeMap<Variable, TypedValue> = BTreeMap::default();
+
+        for computed_table in self.computed_tables.iter() {
+            match computed_table {
+                &ComputedTable::NamedValues { ref names, ref values } => {
+                    for (i, item) in names.iter().enumerate() {
+                        computed_types.insert(item.clone(), values[i].clone());
+                    }
+                },
+                _ => (),
+            }
+        }
+
+        let mut empty_reason: Option<EmptyBecause> = None;
+        let mut wheres_to_add = vec![];
+        for (v, value_type_set) in self.known_types.iter() {
+            if self.column_bindings.contains_key(v) {
+                let cols = self.column_bindings.get(v).unwrap().clone();
+                for col in cols {
+                    match col.1 {
+                        Column::Fixed(DatomsColumn::Value) => { 
+                            // attempt to figure out exactly which type is used
+                            // TODO: But only if the attribute is a variable or not present and therefore of unknown type
+                            let mut value_type: Option<ValueType> = None;
+                            for constraint in self.wheres.0.iter() {
```

Comment:
There has to be a better way of doing this. The CC is our own intermediate representation: if we're having to crawl over it to figure out what's happening, then it means we're not tracking something we should be!